File: basaltpass-frontend/src/shared/api/tenant/crossAppTrust.ts
import client from '../client'

export interface CrossAppTrust {
  id: number
  tenant_id: number
  source_app_id: number
  target_app_id: number
  allowed_scopes: string
  max_token_ttl: number
  description: string
  is_active: boolean
  created_by: number
  created_at: string
  updated_at: string
  source_app?: { id: number; name: string; icon_url: string }
  target_app?: { id: number; name: string; icon_url: string }
}

export interface TokenExchangeLogEntry {
  id: number
  user_id: number
  source_client_id: string
  source_app_id: number
  target_app_id: number
  requested_scopes: string
  granted_scopes: string
  token_ttl: number
  status: 'granted' | 'denied'
  deny_reason?: string
  ip: string
  created_at: string
  user?: { id: number; email: string; nickname: string }
  source_app?: { id: number; name: string }
  target_app?: { id: number; name: string }
}

export interface CreateCrossAppTrustRequest {
  source_app_id: number
  target_app_id: number
  allowed_scopes: string
  max_token_ttl: number
  description?: string
}

export interface UpdateCrossAppTrustRequest {
  allowed_scopes?: string
  max_token_ttl?: number
  description?: string
  is_active?: boolean
}

export const crossAppTrustApi = {
  async list(page = 1, pageSize = 20) {
    const response = await client.get<{
      data: {
        trusts: CrossAppTrust[]
        total: number
        page: number
        page_size: number
      }
    }>('/api/v1/tenant/cross-app-trusts', { params: { page, page_size: pageSize } })
    return response.data.data
  },

  async create(data: CreateCrossAppTrustRequest) {
    const response = await client.post<{ data: CrossAppTrust }>('/api/v1/tenant/cross-app-trusts', data)
    return response.data.data
  },

  async update(id: number, data: UpdateCrossAppTrustRequest) {
    const response = await client.patch<{ data: CrossAppTrust }>(`/api/v1/tenant/cross-app-trusts/${id}`, data)
    return response.data.data
  },

  async delete(id: number) {
    const response = await client.delete<{ message: string }>(`/api/v1/tenant/cross-app-trusts/${id}`)
    return response.data
  },

  async listLogs(params: { page?: number; page_size?: number; status?: string; source_app_id?: number; target_app_id?: number } = {}) {
    const response = await client.get<{
      data: {
        logs: TokenExchangeLogEntry[]
        total: number
        page: number
        page_size: number
      }
    }>('/api/v1/tenant/cross-app-trusts/logs', { params })
    return response.data.data
  },
}

File: basaltpass-backend/internal/model/cross_app_trust.go
package model

import "time"

// CrossAppTrust defines a trust relationship that allows SourceApp to perform
// Token Exchange and obtain a narrowed-scope access token for TargetApp on
// behalf of users.  Managed by tenant admins.
//
// RFC 8693 — OAuth 2.0 Token Exchange
type CrossAppTrust struct {
	ID            uint      `gorm:"primaryKey" json:"id"`
	TenantID      uint      `gorm:"not null;index" json:"tenant_id"`
	SourceAppID   uint      `gorm:"not null;index" json:"source_app_id"`   // the app that initiates token exchange (e.g. WeavInt)
	TargetAppID   uint      `gorm:"not null;index" json:"target_app_id"`   // the app whose API is being called (e.g. Araneae)
	AllowedScopes string    `gorm:"type:text" json:"allowed_scopes"`       // comma-separated scope list that can be granted
	MaxTokenTTL   int       `gorm:"default:300" json:"max_token_ttl"`      // max token lifetime in seconds (default 5 min)
	Description   string    `gorm:"size:500" json:"description"`           // optional human-readable description
	IsActive      bool      `gorm:"default:true" json:"is_active"`
	CreatedBy     uint      `gorm:"not null" json:"created_by"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`

	// Relations
	Tenant    Tenant `gorm:"foreignKey:TenantID" json:"tenant,omitempty"`
	SourceApp App    `gorm:"foreignKey:SourceAppID" json:"source_app,omitempty"`
	TargetApp App    `gorm:"foreignKey:TargetAppID" json:"target_app,omitempty"`
	Creator   User   `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
}

// TableName sets the database table name.
func (CrossAppTrust) TableName() string {
	return "cross_app_trusts"
}

// TokenExchangeLog records every token exchange attempt for auditing.
type TokenExchangeLog struct {
	ID              uint      `gorm:"primaryKey" json:"id"`
	TenantID        uint      `gorm:"not null;index" json:"tenant_id"`
	UserID          uint      `gorm:"not null;index" json:"user_id"`
	SourceClientID  string    `gorm:"size:64;not null;index" json:"source_client_id"`
	SourceAppID     uint      `gorm:"not null" json:"source_app_id"`
	TargetAppID     uint      `gorm:"not null;index" json:"target_app_id"`
	TrustID         uint      `gorm:"index" json:"trust_id"`
	RequestedScopes string    `gorm:"type:text" json:"requested_scopes"`
	GrantedScopes   string    `gorm:"type:text" json:"granted_scopes"`
	TokenTTL        int       `gorm:"not null" json:"token_ttl"`
	Status          string    `gorm:"size:20;not null" json:"status"`        // "granted" or "denied"
	DenyReason      string    `gorm:"size:500" json:"deny_reason,omitempty"` // reason when denied
	IP              string    `gorm:"size:45" json:"ip"`
	CreatedAt       time.Time `json:"created_at"`

	// Display relations (read only)
	User      User `gorm:"foreignKey:UserID" json:"user,omitempty"`
	SourceApp App  `gorm:"foreignKey:SourceAppID" json:"source_app,omitempty"`
	TargetApp App  `gorm:"foreignKey:TargetAppID" json:"target_app,omitempty"`
}

// TableName sets the database table name.
func (TokenExchangeLog) TableName() string {
	return "token_exchange_logs"
}
